### ISSUE TITLE
Time taken and Conditions for finishing training

### ISSUE THREAD

**Issue by Dionysus061726** [Coordi777/Conditional-Diffusion-for-SAR-to-Optical-Image-Translation#10]
Hi! Thanks for your work! I encountered some problems during training that I would like to ask for some help:

I have been training on my dataset for about 10 days, and the training steps have reached 1.91e+06. How can I know if the model has finished training? Or does the training process continue indefinitely if I don’t stop it manually?

In other words, what are the conditions for the training to end? Where is this reflected in the code?

-------------------------
| grad_norm  | 0.0099   |
| loss       | 0.00519  |
| loss_q0    | 0.0155   |
| loss_q1    | 4.72e-05 |
| loss_q2    | 1.71e-05 |
| loss_q3    | 8.83e-06 |
| mse        | 0.00519  |
| mse_q0     | 0.0155   |
| mse_q1     | 4.72e-05 |
| mse_q2     | 1.71e-05 |
| mse_q3     | 8.83e-06 |
| param_norm | 939      |
| samples    | 1.15e+07 |
| step       | 1.91e+06 |
-------------------------

**Comment by Dionysus061726** [comment]
Here is the loss fig

![Figure_1](https://github.com/user-attachments/assets/5675cc6c-2a4b-4f90-b384-2a073cad3a70)

**Comment by szh404** [comment]
> Hi! Thanks for your work! I encountered some problems during training that I would like to ask for some help:
> 
> I have been training on my dataset for about 10 days, and the training steps have reached 1.91e+06. How can I know if the model has finished training? Or does the training process continue indefinitely if I don’t stop it manually?
> 
> In other words, what are the conditions for the training to end? Where is this reflected in the code?
> 
> ## | grad_norm | 0.0099 |
> | loss | 0.00519 |
> | loss_q0 | 0.0155 |
> | loss_q1 | 4.72e-05 |
> | loss_q2 | 1.71e-05 |
> | loss_q3 | 8.83e-06 |
> | mse | 0.00519 |
> | mse_q0 | 0.0155 |
> | mse_q1 | 4.72e-05 |
> | mse_q2 | 1.71e-05 |
> | mse_q3 | 8.83e-06 |
> | param_norm | 939 |
> | samples | 1.15e+07 |
> | step | 1.91e+06 |

Hi
代码不会自动停止，除非你手动结束它
The code will not stop automatically unless you manually stop it.

对应的代码:
guided_diffusion/train_utils.py文件
```
while (
                not self.lr_anneal_steps
                or self.step + self.resume_step < self.lr_anneal_steps
        ):
```

**Comment by Dionysus061726** [comment]
@szh404 Thank you! I didn't read the article carefully where the author specified that his iter steps was 50000. It's my fault. I also found out later that it was the matter of the number of iteration.

 I had no deep learning experience, so I was confused about this. Anyway, I'll keep learning.

**Comment by Yyang-sys** [comment]
你好楼主，我读了文章，但是没搞懂，到底应该训练到什么情况才结束，训练保存的模型在哪里

**Comment by szh404** [comment]
> 你好楼主，我读了文章，但是没搞懂，到底应该训练到什么情况才结束，训练保存的模型在哪里

这个模型训练不会结束，是靠自己手动结束的。如果你担心效果，可以从断点处继续训练，即设置scripts/image_train.py文件夹里的`create_argparser`函数的`resume_checkpoint`参数，比如：`resume_checkpoint="/home/gf/work/szh/Conditional_diffusion/Gaofen3_SAROPT_diffusionstep2000/model180000.pt",`

**Comment by Yyang-sys** [comment]
create_argparserresume_checkpointresume_checkpoint="/home/gf/work/szh/Conditional_diffusion/Gaofen3_SAROPT_diffusionstep2000/model180000.pt",请问这个举例就是手动停止训练后保存的模型路径吗，还是需要在train.sh 中export OPENAI_LOGDIR=/dirs/to/log设置路径

**Comment by szh404** [comment]
> create_argparserresume_checkpointresume_checkpoint="/home/gf/work/szh/Conditional_diffusion/Gaofen3_SAROPT_diffusionstep2000/model180000.pt",请问这个举例就是手动停止训练后保存的模型路径吗，还是需要在train.sh 中export OPENAI_LOGDIR=/dirs/to/log设置路径

这二者不冲突，resume_checkpoint你设置的模型，就是你保存在log中的模型。你需要使用`OPENAI_LOGDIR=./logs`，设置好保存路径后，开始训练，训练到一定阶段，手动停止，进行测试。如果效果没有达到要求，可以选择从./logs里找出保存的模型接着训练。大体是这么个意思

**Comment by Yyang-sys** [comment]
明白了，十分感谢！！best wishes

**Comment by Yyang-sys** [comment]
> > create_argparserresume_checkpointresume_checkpoint="/home/gf/work/szh/Conditional_diffusion/Gaofen3_SAROPT_diffusionstep2000/model180000.pt",请问这个举例就是手动停止训练后保存的模型路径吗，还是需要在train.sh 中export OPENAI_LOGDIR=/dirs/to/log设置路径
> 
> 这二者不冲突，resume_checkpoint你设置的模型，就是你保存在log中的模型。你需要使用`OPENAI_LOGDIR=./logs`，设置好保存路径后，开始训练，训练到一定阶段，手动停止，进行测试。如果效果没有达到要求，可以选择从./logs里找出保存的模型接着训练。大体是这么个意思

你们，抱歉再次打扰。请问测试的时候加入我有”model00000X0000.pt“模型文件，测试的时候是否把启动测试命令中”python scripts/image_sample_realtime.py --model_path /path/to/checkpoints $MODEL_FLAGS $DIFFUSION_FLAGS $SAMPLE_FLAGS“中的”model_path /path/to/checkpoints“改成”model_path ./logs/model00000X0000.pt“即可

**Comment by Yyang-sys** [comment]
> > > create_argparserresume_checkpointresume_checkpoint="/home/gf/work/szh/Conditional_diffusion/Gaofen3_SAROPT_diffusionstep2000/model180000.pt",请问这个举例就是手动停止训练后保存的模型路径吗，还是需要在train.sh 中export OPENAI_LOGDIR=/dirs/to/log设置路径
> > 
> > 
> > 这二者不冲突，resume_checkpoint你设置的模型，就是你保存在log中的模型。你需要使用`OPENAI_LOGDIR=./logs`，设置好保存路径后，开始训练，训练到一定阶段，手动停止，进行测试。如果效果没有达到要求，可以选择从./logs里找出保存的模型接着训练。大体是这么个意思
> 
> 你们，抱歉再次打扰。请问测试的时候加入我有”model00000X0000.pt“模型文件，测试的时候是否把启动测试命令中”python scripts/image_sample_realtime.py --model_path /path/to/checkpoints $MODEL_FLAGS $DIFFUSION_FLAGS $SAMPLE_FLAGS“中的”model_path /path/to/checkpoints“改成”model_path ./logs/model00000X0000.pt“即可

还是说指向“ema_0.99_xxxxxx.pt”文件

**Comment by szh404** [comment]
> > > > create_argparserresume_checkpointresume_checkpoint="/home/gf/work/szh/Conditional_diffusion/Gaofen3_SAROPT_diffusionstep2000/model180000.pt",请问这个举例就是手动停止训练后保存的模型路径吗，还是需要在train.sh 中export OPENAI_LOGDIR=/dirs/to/log设置路径
> > > 
> > > 
> > > 这二者不冲突，resume_checkpoint你设置的模型，就是你保存在log中的模型。你需要使用`OPENAI_LOGDIR=./logs`，设置好保存路径后，开始训练，训练到一定阶段，手动停止，进行测试。如果效果没有达到要求，可以选择从./logs里找出保存的模型接着训练。大体是这么个意思
> > 
> > 
> > 你们，抱歉再次打扰。请问测试的时候加入我有”model00000X0000.pt“模型文件，测试的时候是否把启动测试命令中”python scripts/image_sample_realtime.py --model_path /path/to/checkpoints $MODEL_FLAGS $DIFFUSION_FLAGS $SAMPLE_FLAGS“中的”model_path /path/to/checkpoints“改成”model_path ./logs/model00000X0000.pt“即可
> 
> 还是说指向“ema_0.99_xxxxxx.pt”文件

指数移动平均（EMA）是指最近训练步骤的平均权重，而不是最后一个训练步骤的权重（我查的），采用ema算法的效果更好。你测试可以用带ema来进行测试。

**Comment by Yyang-sys** [comment]
> > > > > create_argparserresume_checkpointresume_checkpoint=“/home/gf/work/szh/Conditional_diffusion/Gaofen3_SAROPT_diffusionstep2000/model180000.pt”，请问这个举例就是手动停止训练后保存的模型路径吗，还是需要在 train.sh 中export OPENAI_LOGDIR=/dirs/to/log设置路径
> > > > 
> > > > 
> > > > 这二者不冲突，resume_checkpoint你设置的模型，就是你保存在log中的模型。你需要使用，设置好保存路径后，开始训练，训练到一定阶段，手动停止，进行测试。如果效果没有达到要求，可以选择从./logs里找出保存的模型接着训练。大体是这么个意思`OPENAI_LOGDIR=./logs`
> > > 
> > > 
> > > 你们，抱歉再次打扰。请问测试的时候加入我有“model00000X0000.pt”模型文件，测试的时候是否把启动测试命令中“python scripts/image_sample_realtime.py --model_path /path/to/checkpoints $MODEL_FLAGS $DIFFUSION_FLAGS $SAMPLE_FLAGS”中的“model_path /path/to/checkpoints”改成“model_path ./logs/model00000X0000.pt”即可
> > 
> > 
> > 还是说指向“ema_0.99_xxxxxx.pt”文件
> 
> 指数移动平均（EMA）是指最近训练步骤的平均权重，而不是最后一个训练步骤的权重（我查的），采用ema算法的效果更好。你测试可以用带ema来进行测试。

你好，他的论文中，FID和SSIM的指标是怎样跑出来的，项目中好像没有提供

**Comment by Mizar29** [comment]
@Dionysus061726 Hi, how is the image quality of your reproduction? The image quality of my reproduction is very poor.
